OpenLiberty/open-liberty
# Issue 16167: Fixes and Kerberos Ldap FAT support

Fixes #16167

Two runtime changes, both discussed in slack:

1. Added a doPriv in the login module

2. Added a clear option for the KerberosLRU cache for LdapRegistry/ContextManager to call

Initial check in of FAT infrastructure and some FAT tests. I have some simple bind and ticketCache tests. Keytab is still in progress. Still have some to-dos to address -- follow up #16181

General structure:

-  the ApacheDS infrastracture is started in a beforeClass/afterClass globally for the bucket.

- There's a general helper class.

- There's a base UserLoginTest that the different bind type tests will expand on. It has a baseline "login" test where I do a good/bad password check of a user, isValidUser, getGroupsForUser and getUsersForGroup.

Function covered so far:
`SimpleBindTest` -- running baseline login
- BindAuthMech is simple
- BindAuthMech is none, Ldap allows anon bind and disallows
- BindAuthMech is not set
- `swapFromSimpleToKerberos` is in progress

`TicketCacheBindTest`
- Cannot contact KDC
- Empty, missing, or invalid `krb5PrincipalName`
- Empty, missing, or read-only  `krb5TicketCache`
- Start with expired cache, manually updated to valid
- Swap to keytab

`TicketCacheBindMultiRegistryTest`
- Use two Ldaps (ApacheDS and UnboundID) with allowOp = false. Confirm that neither registry can login when ApacheDS is down or cache is expired (and recovers when Apacheds is up, valid cache).
- Use two Ldaps (ApacheDS and UnboundID) with allowOp = true. Confirm that UnboundID can login when ApacheDS is down or cache is expired (and recovers when Apacheds is up, valid cache).

`TicketCacheBindLongRunTest`
- Restart ApacheDS server, confirm login recovers after LDAP restart
- Start with a valid certificate that is renewable and expires during runtime, verify ticket is renewed
- Start with a valid certificate that is not renewable and expires during runtime, ticket is not renewed, but manual renew works.
